### ISSUE TITLE
Energy management/calculations

### ISSUE THREAD

**Issue by jlaur** [openhab/openhab-core#3478]

**Comment by jlaur** [comment]
@kaikreuzer - great to see some concrete proposals for the architecture, thanks! Since I have mostly been working within the price (including calculations) domain so far, I would be interested into digging a bit more into the `EnergyProvider` interface. My only power source is the grid, so more specifically `getPrice(interval)`.

Do you imagine `EnergyManagementService` requesting all prices within a given interval, so that it can perform calculations, such as the cost of a given consumption within a given period, or do you expect each `EnergyProvider` implementation to perform such calculations? Having different sources of prices (bindings implementing `EnergyProvider`) with the same "calculation engine" seems desirable to avoid duplication? In this case the provider would either need to provide a sum of the price elements for the available intervals, or provide the elements themselves.

In the openhab/openhab-addons#14376 I decided to provide the elements individually and let the user create a `Group` item or similar for calculating the sum: [channels](https://github.com/openhab/openhab-addons/blob/2581fc3855dfd3743355262885cecf1ea3c4cf6f/bundles/org.openhab.binding.energidataservice/README.md#channel-group-electricity). This provides a certain level of flexibility as a base price can be added without the binding needing to support this, or elements could even come from other external sources. For example, one binding could provide the spot price and another one the relevant tariffs for a specific grid company or area.

So I'm thinking, if the `EnergyManagementService` should perform the calculations, some degree of configuration is probably needed for picking relevant sources. Perhaps each implementation could expose which kind of elements it supports (e.g. `energidataservice:spot-price`), so that this could be requested specifically when asked to provide prices for a given interval.

Calculations can be done without VAT internally to find lowest relative price etc. or the engine could support transformations, e.g. openhab/openhab-addons#14529

A few additional unstructured thoughts for which I currently have no idea where they fit in (or if they do at all).
- Energy is not only electricity, it could also be gas. Energi Data Service also has datasets for that.
- In Denmark you have to pay net tariffs when delivering electricity (i.e. from PV) to the grid. When spot prices are low, you might have to pay more than you receive for delivering. From a pure economical perspective, this could incentivize to use excess power by any means.
- Energi Data Service also provides datasets related to CO2 emissions and how "green" (renewable) the electricity will be at given hours. This does not necessarily map directly to the prices. So if one would prioritize Planet Earth over price, this data could be a driver for decision-making rather than prices.

**Comment by kaikreuzer** [comment]
> Do you imagine EnergyManagementService requesting all prices within a given interval,

Yes, I would think it would be simplest, if the EMS regularly scrapes the prices from the providers - possibly up to every minute. In a future iteration, we could also think about a push/notification mechanism upon changes, but I'd think it should be fine without.

> Having different sources of prices (bindings implementing EnergyProvider) with the same "calculation engine" seems desirable to avoid duplication? 

Yes, absolutely. We could have one implementation of an "GridEnergyProvider", which takes the prices from an item that is configurable and which could be filled by one of many bindings.
Other providers, such as a PV system could instead simply deliver a fixed number as a price and wouldn't need such a flexibility.

> In the https://github.com/openhab/openhab-addons/pull/14376 I decided to provide the elements individually

This is fine. If the `hourly-prices` would then be replaced by a Number channel where the prices are written to with a future timestamp (with #3000), this could be easily used for charts and also be consumed by an `GridEnergyProvider` - if we manage to implement this class in a configurable and generic way, we could possibly add this as a default implementation to the core directly.

> one binding could provide the spot price and another one the relevant tariffs for a specific grid company or area.

Yes, absolutely - that could all be configuration options for this provider.

> Calculations can be done without VAT internally to find lowest relative price etc. or the engine could support transformations,

Yeah, here I am not yet really clear - as just mentioned, it could be nice if that grid provider became part of the core, but we shouldn't have any dependency on an add-on such as the VAT transformation... We'll have to think about how to solve this.

> In Denmark you have to pay net tariffs when delivering electricity (i.e. from PV) to the grid.

Even for the case that you are paid money for it, we would need a way to express this - this is currently not covered. Maybe it could be another configuration to the EMS, so that it knows how to consider excess power in its calculations.

**Comment by masipila** [comment]
> 
> > Having different sources of prices (bindings implementing EnergyProvider) with the same "calculation engine" seems desirable to avoid duplication?
> 
> Yes, absolutely. We could have one implementation of an "GridEnergyProvider", which takes the prices from an item that is configurable and which could be filled by one of many bindings. Other providers, such as a PV system could instead simply deliver a fixed number as a price and wouldn't need such a flexibility.

Just to make sure that I follow correctly:

- There could be a Binding called Entso-E Market Prices that would implement EnergyProvider. This would be responsible for fetching the day-ahead market prices from Entso-E API. 
- Alternatively there could be another binding like Tibber or whatever that would fetch similar data from some other API.
- I could contribute another binding called Caruna Tariff Calculator. There is no API where the data could be fetched but it's irrelevant where the source data comes from as long as the Binding is able to produce the desired result.
- And somebody else could contribute a Binding that would fetch / calculate the tariffs for other network companies

> 
> > In the [openhab/openhab-addons#14376](https://github.com/openhab/openhab-addons/pull/14376) I decided to provide the elements individually
> 
> This is fine. If the `hourly-prices` would then be replaced by a Number channel where the prices are written to with a future timestamp (with #3000), this could be easily used for charts

Hooray!

> > Calculations can be done without VAT internally to find lowest relative price etc. or the engine could support transformations,
> 
> Yeah, here I am not yet really clear - as just mentioned, it could be nice if that grid provider became part of the core, but we shouldn't have any dependency on an add-on such as the VAT transformation... We'll have to think about how to solve this.

Many market price providers provide the data in EUR / MWh without VAT whereas most consumers are used to c / kWh as the unit of measure and with VAT. So I would really like to see how the user could have the divide by ten (to convert from EUR / MWh to c / kWh) and multiple by VAT multiplier. It could even be used as "close enough" currency exchange calculations where a Swedish / Danish / Norwegian user could fetch the data from Entso-E API (which responds with EUR / MWh) and convert it with a fixed multiplier to krones / kWh.   

![image](https://user-images.githubusercontent.com/20110757/227443732-0c88ca1e-8651-442b-b816-95ed0458588a.png)

If the attribute 'granularity' here refers to the time resolution (every 15 mins, every 60 mins, every whatever), can we please call this as timeResolution instead of 'granularity'? The term resolution has been used in all industry documents that I have seen.

**Comment by kaikreuzer** [comment]
> Just to make sure that I follow correctly:

Yes, we could have many bindings that each bring their own `EnergyProvider` and which have a channel through which they provide future energy prices to a linked item. Such bindings would have to deal with VAT, currencies, etc. on their own, though.

> So I would really like to see how the user could have the divide by ten

Maybe we could additionally have a generic `GridEnergyProvider` for the "standard" cases, which takes a "future prices" items as a configuration and which offers additional features such as adding VAT, converting from EUR to ct or adding another channel for fixed costs/taxes into the calculation.

> can we please call this as timeResolution instead of 'granularity'? 

Sure thing.

**Comment by J-N-K** [comment]
I would be willing to look into the currency issue, there is already #3408 for that. I would like to hear some comments if we should try to fit it into UoM or create a new type for that. 

Another thing (which is probably not so prominent like electricity): there may also be dynamic water or gas tariffs that should be included and a consumer could use more than one of them (e.g. dish washer).

**Comment by kaikreuzer** [comment]
Thanks @J-N-K. My first impuls is that it should be fitted into UoM, but I haven't thought through the consequences (mainly wrt dynamic exchange rates).

> a consumer could use more than one of them (e.g. dish washer).

That's true, but are there really any countries in the world where there are dynamic tariffs for water? I think adding such a feature would heavily increase complexity as the search for optima would have to be done for different dimensions simultaneously - I'd rather prefer to stay away from this (for now).

**Comment by J-N-K** [comment]
I had a quick look and I think we have to extend `AbstractQuantity` to a new `MoneyQuantity`. We could then implement`.toUnit` to use a `CurrencyConversionProvider`. It's not super easy, but I think it's doable. The benefit is that we don't need a new data type for that and can easily do any calculations we want, even besides this topic (e.g. calculate the cost of garden watering based on consumption.

I'm not aware of any country that does it. I would not implement that in the first step, just if it's possible to keep it extendable then we should do it. We might see gas tariffs in the future that at least have daily or monthly changes (especially thinking of hydrogen replacing fossil gas).

**Comment by jlaur** [comment]
> > In the [openhab/openhab-addons#14376](https://github.com/openhab/openhab-addons/pull/14376) I decided to provide the elements individually

Just to be clear, in lack of a better word, with (price) "elements" I mean the components that make up the total price you have to pay per kWh. In my case this is spot price, net tariff, system tariff, electricity tax and transmission net tariff. On top of that the fee to my electricity, which is luckily fixed (no binding for that 🙂).

> This is fine. If the `hourly-prices` would then be replaced by a Number channel where the prices are written to with a future timestamp (with #3000), this could be easily used for charts and also be consumed by an `GridEnergyProvider` - if we manage to implement this class in a configurable and generic way, we could possibly add this as a default implementation to the core directly.

👍

> > one binding could provide the spot price and another one the relevant tariffs for a specific grid company or area.
> 
> Yes, absolutely - that could all be configuration options for this provider.

So, as an example - two of my price components (let's try that word now) are dynamic: Spot price and net tariff.

Currently both are provided by Energi Data Service, but let's say I wanted to use another binding (or a rule, script calling the REST API, karaf command etc.) for getting the spot price. I would now have at least two different price components (items) from different sources that should be taken into consideration when performing calculations, as the cheapest period has to be calculated from the sum.

Maybe you already had this in mind, so just to be sure and fully aligned. The user should be able to configure which items to include in the calculations.

**Comment by masipila** [comment]
@jlaur that's clear to me. For me, I can't get the grid tariff and the spot price from the same datasource, but that should not be an issue. I could have two separate Bindings, on which fetches the spot prices from Entso-E and another that calculates the grid tariffs and electricity tax.

According to this [1], if the user places the different items in the same group and persists the group, then the sum of all items in that group will be persisted and we have the total price. But doing the summing should not be difficult even if it didn't happen automagically by persisting the group.

If @kaikreuzer 's proposed class diagram is OK for everyone, should we break this thing into separate issues and establish a backlog of them?

1. One issue would be to get the Binding needs to be able to persist data with future timestamps, and I believe @kaikreuzer was in favor of covering this in the scope of https://github.com/openhab/openhab-core/pull/3000

2. If we start by developing a capability to fetch and store the prices, we would next need EnergyManagementParticipantRegistry, EnergyManagementParticipantProvider,
EnergyManagementParticipant and EnergyProvider

3. Once we have 1 and 2, we should be able to create the first Bindings for different price data sources, for example Entso-E, the Danish EnergiDataService and so on.

4. We could then proceed to EnergyConsumer and PowerProfile

Does this sound like a plan?

[1] https://community.openhab.org/t/energy-rules-sum-of-group-to-item-help/10264/5

**Comment by lsiepel** [comment]
> @jlaur that's clear to me. For me, I can't get the grid tariff and the spot price from the same datasource, but that should not be an issue. I could have two separate Bindings, on which fetches the spot prices from Entso-E and another that calculates the grid tariffs and electricity tax.
> 
> According to this [1], if the user places the different items in the same group and persists the group, then the sum of all items in that group will be persisted and we have the total price. But doing the summing should not be difficult even if it didn't happen automagically by persisting the group.
> 
> If @kaikreuzer 's proposed class diagram is OK for everyone, should we break this thing into separate issues and establish a backlog of them?
> 
> 1. One issue would be to get the Binding needs to be able to persist data with future timestamps, and I believe @kaikreuzer was in favor of covering this in the scope of [Enable binding to store historic states #3000](https://github.com/openhab/openhab-core/pull/3000)
> 2. If we start by developing a capability to fetch and store the prices, we would next need EnergyManagementParticipantRegistry, EnergyManagementParticipantProvider,
>    EnergyManagementParticipant and EnergyProvider
> 3. Once we have 1 and 2, we should be able to create the first Bindings for different price data sources, for example Entso-E, the Danish EnergiDataService and so on.
> 4. We could then proceed to EnergyConsumer and PowerProfile
> 
> Does this sound like a plan?
> 
> [1] https://community.openhab.org/t/energy-rules-sum-of-group-to-item-help/10264/5

With persisting, do you mean storing it in some database? As i don't see that as a requirement, the `EnergyManagementService` could keep data in memory. 

Anyway, i see some kind of gap between the proposed `items` (openHAB architecture) and the data structure (requirements), but maybe that is a limitation of my head :-)  
Atleeast future prices and the powerprofile consist of many datapoints depending on how far to look ahead and the time resolution. These two datatypes may consist of 100+ datapoints each. How would these items be structured in a maintainable and understandable way for end-users?

I would have hoped that the framework could take care _direct_ communcation in the background between the `EnergyProvider`, `EnergyConsumer` and the `EnergyManagement` service.

**Comment by kaikreuzer** [comment]
> The user should be able to configure which items to include in the calculations.

@jlaur Yes, agreed.

> Does this sound like a plan?

@masipila Yes, sounds good to me.
For (1), I have just commented on #3000.
For (2), I'll try to come up with a suggestion / draft PR soon®.

> I would have hoped that the framework could take care direct communcation in the background between the EnergyProvider, EnergyConsumer and the EnergyManagement service.

@lsiepel Yes, this would be exactly my plan as well - there can only be specific `EnergyProvider`s that can consume persisted future item states (i.e. future prices) from items as well. PowerProfiles will rather be always provided through a direct interface to the EMS.

**Comment by kaikreuzer** [comment]
Hey guys, I just want to say that I am still alive and haven't forgotten about this topic. I'll probably not be able to make big advancements before the OH4 release end of June, but it is definitely my absolutely top priority right after that.

**Comment by wborn** [comment]
It would also be nice if the solution would be a bit more generic. For instance you might also want to have something for redistributing rain water, money, heat, cold etc. one day. 🙂

**Comment by subbamaggus** [comment]
hi,

i went through the whole issue here.
is there already a start in the core? is it possible to start it as add on?

one thing i would like to add here:
the feature here is mostly talking about planning the future. how about a judgement for the moment?
the simplest one could be "i have more power than i need, store it". for this there is no need for storing a forecast on price or weather or sth more. sure that should be added later.... but first things first, right?

i will checkout the main branch and see if i could find some work done on this.

sounds like fun!!!

**Comment by andrewfg** [comment]
I think somebody is already working on a UoM for converting money amounts i.e USD per Euro (say). And (part of) your discussion is very similar to that -- namely kWh per Euro (say) which you might convert via UoM to Joule per DKr (say). Both these would be extensions of UoM where the conversion factors are variable rather than fixed in code. The variable rates are a time series where past rates are historical actual values, and future rates are forecasts. The source of these rates might be e.g. Bank of Denmark (say) or EON (say), or in my latter example (with double UoM conversion) both sources (say). The intervals between two fixed rate UoM conversion data points could be subject to various interpolation algorithms (e.g. linear, or keep last value, ..)

**Comment by mstormi** [comment]
Now with OH4(.1) out and [Number:Currency and Number:EnergyPrice implemented](https://community.openhab.org/t/using-the-new-currency-units-of-measurement-in-4-1/152338), shouldn't we be picking this up again ?

I just applied for a grid discount according to German §14a EnWG and my provider responded they currently are thinking to [implement EEBus as a control protocol to power-dim or shutoff large consumers](https://community.openhab.org/t/searching-for-willing-to-create-a-general-energy-control-binding-add-on/149165/5). Just spreading the word. 

I believe openHAB would benefit from an EEBus implementation.
Anyone willing to take on the challenge ?

**Comment by florian-h05** [comment]
There is a community developed EEBus binding: https://github.com/csowada/openhab-ebus-binding.
Probably it should be attempted to make it a official one.

**Comment by splatch** [comment]
@florian-h05 Binding you refer to is ebus, while Markus talks about eebus (https://www.eebus.org/). I know, its quite confusing, but later is more recent and (in theory) spans across more areas than just heating/ventilation.

**Comment by florian-h05** [comment]
Oh thanks for the hint, seems like it also confused me ;-)

**Comment by mstormi** [comment]
For what's it worth, EVCC has a working EEBus implementation in Go and [is now working on v2](https://github.com/evcc-io/evcc/pull/12407).

On forecasting, what about finalizing https://github.com/openhab/openhab-addons/pull/13308 for 4.2.
@jlaur @weymann what's missing ?

**Comment by jlaur** [comment]
> On forecasting, what about finalizing [openhab/openhab-addons#13308](https://github.com/openhab/openhab-addons/pull/13308) for 4.2.
> @jlaur @weymann what's missing ?

The current status can be seen in the PR - last comment.

**Comment by DerAndereAndi** [comment]
I would like to make some comments about EEBUS:

- ebus and EEBUS are two completely different things, they have nothing in common
- the evcc project uses the EEBUS implementation of https://github.com/enbility/ , which I am developing. It consists of 4 repositories fully under MIT license and is a major task to implement. I am working on them since 2 years and had developed a prototypical version over 2 years before that. It is impossible to implement it within a few months or weeks. It is way too complex.
- The [CEMd repository's](https://github.com/enbility/cemd) goal is to provide a a simple API so EEBUS can be used with only little knowledge about the protocol itself. One could create a service on top of it which then e.g. could use a websocket interface to other systems like OpenHab which are not based on Go.
- During the last year I am improving an refactoring the repositories a lot and target a 0.5 release in the near future. This is what evcc is calling V2.

Integration would still be a big task, as the service on top of CEMd would have to be written in Go and an interface would have to be defined. Once that is done, support for germans §14a technically could also be implemented. Feel free to get in touch, my contact details are available at https://enbility.net

**Comment by florian-h05** [comment]
I guess a different option would be to write a JNI wrapper for your EEBus implementation.

I have googled a bit, and I also found the Java EEBus implementation from openMUC, which is created by a team of the Fraunhofer ISE in Freiburg, Germany: https://www.openmuc.org/eebus/

**Comment by DerAndereAndi** [comment]
@florian-h05 sadly it is not open source and I haven’t seen it working with a device yet. But maybe they would be interested anyway helping you to have a show case

**Comment by kaikreuzer** [comment]
I guess I have to face the truth of my situation that I won't be able to find the time to work on that topic the way I intended. Since I spent some time and [proposed a potential architecture](https://github.com/openhab/openhab-core/issues/3478#issuecomment-1481931374) for tackling the problem, exactly one year has passed without any progress on my end - I'm feeling very sorry about that. 😢 

So please do not wait for me here - if anybody is interested to pick up this topic, please speak up and drive it forward!
Wrt the EEBus discussion: As @DerAndereAndi rightly says, I don't see a native EEBus implementation in openHAB feasible; I am not even convinced that a proper EEBus implementation is possible in the long-run, since my impression/experience is that this is a very commercially-focussed initiative that rather tries to keep control and which makes it hard for open-source to get a grip on it.

In any case, I would split the discussion about general energy management (this issue) and EEBus. Imho, if we are able to create a modular architecture like the one suggested above, EEBus could possibly be one implementation behind the interfaces, but it would leave the room for many other (simpler) approaches as well.

**Comment by rolfth** [comment]
https://github.com/masipila/openhab-spot-price-optimizer/wiki

**Comment by stamateviorel** [comment]
i read through the whole thread, including @kaikreuzer's architecture proposal and the march 2024 message about not waiting on it. so instead of a +1 i wanted to add something that might actually be useful.

for a while now i have been running a fairly complete energy management on my own openhab, in production — 25 kWp solar + battery, 4 ev chargers, knx and submetering. over time it grew into a binding that does:

- solar forecast (open-meteo, with inverter clipping)
- dynamic tariff + comparing providers
- capacity tariff peak shaving (belgian capaciteitstarief)
- solar first ev charging (eco vs fast)
- cheapest hour boiler / dhw heating
- cost / co2 / self consumption tracking
- a small energy page in mainui on top

the reason i bring it up here: when i built it i ended up with more or less the same shapes @kaikreuzer drew in this issue. there is an energy provider (grid price, pv, battery), an energy consumer with a power profile (the wallbox and the battery are the controllable ones, the boiler is a simple on/off load), and i mark the relevant items with metadata so the engine knows which ones to use. so from the practical side the architecture holds up — i can confirm it works because i basically built a less clean version of it for real.

i am also already leaning on the pieces that landed since 2023, Number:EnergyPrice and the timeseries future states for the day ahead prices and the solar forecast, which are exactly the building blocks that were missing back then.

i know this overlaps with the evcc binding and with @masipila's spot price optimizer, both very good. the parts mine adds on top are the capacity tariff peak shaving, the boiler/dhw cheapest hour, and the native openhab cost/co2/self consumption, and it runs fully inside openhab without an external service.

i am not trying to say make my thing core, it is too tied to my own house. but the architecture in this issue is real and buildable, i have a working reference, and i would be happy to help pick it up — even a minimal EnergyManagementService + the energy: metadata namespace as a start. is there still appetite from the people active here to move it forward? i can share the code, write up how the provider / consumer / profile mapping works in practice, or just help where it is useful. i just did not want a real working implementation to sit unused while this issue is stuck.

**Comment by florian-h05** [comment]
That sounds very interesting.
Do you have your code on GitHub? I’m currently lacking time to look into such “major” things but still interested to do so later.

**Comment by lsiepel** [comment]
This still has my interest, but lack condensed time to start work on an effort like this.

**Comment by mstormi** [comment]
I'm interested, too, but I, too have been lacking time to embark on this.

In case you don't know: I've actually built a (commercial) EMS with openHAB rules. So I _sort of_ have done this already in my implementation.
No free GitHub access but I can share the main code as a OH config backup: https://storm.house/download/ems_demo.zip 

A word of warning, though: we have integrated various providers (pv, bat, dyn. tariff) and consumers (heat pump, ev charger, white goods ...), but really every single one of them differs from all the others _by nature_ and it is really challenging to come up with a comprehensive taxonomy that covers all of their characteristics.

As @jlaur opened this issue with a dish washer, I'll take that as an example for what I mean by that:
A dish washer you cannot switch on/off at any time or even tell it to change the power it is washing with. Or, with a plug in front or smart dishwasher you can, but you shouldn't interrupt the washing for more than some minutes (heat dissipating -> energy loss) and/or for comfort reasons (if you cannot predict any finishing time when the washing will be done, people won't like this).
A dishwasher uses some 10 watts for 10 mins, then 2 kW for some minutes then again some (negligible) 10 watts. That you cannot change. In comparison, most EV chargers you can continously tell every some seconds to change to 2 kW or 11 kW or 0 W, depending on your energy supply or cost (no PV surplus at night, of course).
This just being one example to illustrate the fundamental challenge, there's many more differences among devices and use cases and heat pumps are a story of their own. Some have individual APIs that you can tell to consume a specific amount of power, some even have EEBus (no option, see above) and they all have SGready, but using that you cannot determine the power it's to use.
Actually my HEMS tracks the "energy availability/cost" level across providers (pv, bat, current dyn. price) and maps that into SGready modes, more or less.

This is the true challenge in this issue here and probably reason why noone has picked it up so far:
To create a taxonomy that covers and integrates all types of consumer devices **and providers** and their characteristics in a meaningful, universal way is hard, really hard, and requires a lot of knowledge about all those consumer devices and characteristics out there. (and not just those you use yourself)

**Comment by lsiepel** [comment]
Long time ago i had some thoughts about cases like you describe. I might have even written it down here somewhere. I think the energy usage pattern for a given device can be normalized up to some extend. Providing more and more parameters will make it more and more optimal, but it comes at the cost of complexity. At some point there will be a balance.

Regarding the power usage pattern itself, I can imagine that only a small number of parameters would already allow many devices to fit a common model, making it practical to start building and testing such a system.

For example, a few basic parameters could be enough:

Whether the device follows a fixed or variable power usage pattern.
Whether the device can be paused and resumed without affecting its operation.
A description of its power profile, such as a list of peak power levels and their durations (or an equivalent normalized representation).

With just these characteristics, it should already be possible to make meaningful scheduling decisions for a large class of appliances. The model would not be perfect, but it would provide a common abstraction that works well enough to begin optimizing energy usage. More detailed parameters could always be added later for devices that require a more accurate representation, without complicating the initial implementation for every device.

**Comment by mstormi** [comment]
Right, it's all about finding the right balance in representation accuracy vs. programming complexity.
See my code for my preferred solution and believe me, this is result of an optimization alongside multiple years I've already been spending on this topic so far.
That complexity and differing views on what is important and what is not really is the key challenge and reason why noone picked on this task so far.

**Comment by stamateviorel** [comment]
@florian-h05 yes, it is public: https://github.com/stamateviorel/openhab-binding-emsmanager — the whole binding is there. the part that maps to this issue lives in `src/main/java/org/openhab/binding/emsmanager/internal/ems/` (a small EnergyManagementService, the `energy:` item metadata parsing and a scanner that discovers the tagged providers/consumers from the registry), with a write-up in `docs/EMS_ENGINE.md` of how the provider / consumer / profile mapping works in practice. everything around it is my own house logic — take it as the testbed it runs in, not as a proposal. fair warning: the runner glue in there still carries my testbed/parity scaffolding — the clean parts are the metadata parser, the participant records and the service. and to be clear, in core this would want real interfaces a binding can implement too (kai's original sketch); my metadata-only version is the user-tagging half of it, proven to work.

@mstormi i did not want to just say "will look", so i downloaded your demo backup and read through it properly (Energiemanagement.rules, the tariff and wp STORE trees, weisseWare). first: respect. the vendor STORE alone — about twenty pv vendors, the battery and meter trees, five heat pump families all normalized into one item set — is exactly the "you need to know more devices than just your own" problem, solved by hand. i understand your warning much better now.

what i find really interesting is where our two systems, built independently, ended up the same and where not. same: demand plus a latest-start/deadline per consumer, surplus-driven starts, cheapest and most-expensive hour windows (your bestMaximal / bestVerstaerkt / teuerstSperren is the same shape as my cheapest-hours ranking), a battery reserve floor, priority ordering. different: you plan against an energy balance over the whole horizon — battery above reserve plus forecast remaining yield versus remaining needs until next sunrise — and admit consumers against that budget, while i dispatch more reactively per tick against live surplus. your ev control goes through evcc modes where mine talks ocpp amps directly. and your consumers are richer where mine are thin: min/max runtime and cooldown on the heizstab stages, the sg-ready 4-state for the heat pump, the white goods as start-once machines with a latest-start.

reading both codebases side by side, the taxonomy feels more tractable to me than i feared — not because the differences markus describes are small, but because they are real and they repeat: every consumer in his system and mine falls into one of four profile classes, and two of the four exist precisely because of the differences he warns about:

1. simple on/off (boiler, immersion heater stage) — plus min/max runtime and cooldown as protection parameters
2. continuously controllable (ev charger, battery) — min/max watts or amps, ramp limit
3. mode-controllable (sg-ready heat pump) — a handful of discrete states you can suggest, but no direct power setpoint
4. batch / fixed program (dishwasher — @jlaur's original case at the top of this issue — washing machine) — you only choose the start moment before a latest-start, then it runs its own profile untouched

plus three cross-cutting things that apply to any class: demand (kWh) with a deadline; metering delivered energy against that demand (my hard-learned one — a planner without it keeps heating water that is already hot); and grid-side constraints as first-class inputs (markus handles the german §14a dimming signal, i handle the belgian capacity tariff — different countries, same shape: the grid can cap you).

@lsiepel i think this is your minimal parameter list, just grouped: fixed-vs-variable pattern and pausable-or-not fall out of the class, and the detailed power profile only really matters for class 4. and it is not speculative — classes 1, 2 and 4 plus sg-ready run in markus' production today, classes 1 and 2 plus metered demand run in mine.

i am happy to write this up properly as a github discussion or a small design doc for the `energy:` metadata namespace when one of you has the window to review it. no rush from my side — the code is there to poke at in the meantime.

**Comment by mstormi** [comment]
Thanks @stamateviorel for dissecting my code, well done!

From my experience in practice, I fear we somehow need to factor in another category of devices: dynamic ones, i.e. those that have a control intelligence of their own. Not sure how to handle that, though.
Heat pumps are particularly nasty, they decide for themselves when and how much they heat, even when you control them with SGready.
Battery charging is usually inverter-controlled and some ev chargers do pv surplus charging working on their own sensor's or even internet data input basis, not our openHAB's logic we put on top. Device behavior is always there by default. You might be able to override it from OH but not every aspect of it. It's influenceable via device specific modes, logic and limit parameters but never fully controllable.
And what you can ultimately do and even more importantly so, what you have to pay attention for, depends a lot on device implementations.

Another open question is if and if so how to represent consumption profiles. 
Just an idea: can we represent a prototypical consumption chart in form of an abstract `TimeSeries` of `%` unit type?
Plus a scale factor, plus min and max limits, eventually.
Would do fine for the dish washer example.
In a more generic representation form, this could also be applied to a group of consumers like a household.
In Germany, there's load profiles that are used in official metering and tariffing. You can even download those from the regulator.

Finally, we need to think about profiles on the provider side, too.
If the generation side profile is represented as a `TimeSeries`, too, and dynamically changeable, we could for example use AI to predict PV generation for the day(s) to come, or reuse others' predictions and 'feed' them into our calculation via these profiles.

**Comment by stamateviorel** [comment]
thanks markus — and yes, you just named the thing my four classes quietly ignore. let me try to place it, because i run into it daily.

i don't think "dynamic" is a fifth class next to the other four, because the four describe the control *surface* a device gives you (on/off, watts, modes, start). what you describe is how the device behaves *inside* that surface — it has its own inner control loop. and from running this live my conclusion is: that is not the exception, it is the normal case, and the model should assume it everywhere. every command in classes 2–4 is really an **envelope**, not an order. my wallbox is the clearest example — ocpp SetChargingProfile sets a *limit*, and the car decides what it actually draws inside it (sometimes nothing: SuspendedEV is literally the car declining). so my "controllable" ev charger is honestly a "limitable" one. and i have to laugh a bit here, because my own battery turns out to be the purest proof of your point: while writing this reply i double checked my setup, and the inverter on my site does not accept setpoints at all — my battery "dispatcher" has been computing a time-of-use plan that is advisory only, while the inverter runs its own self-consumption logic and ignores everybody. a class-2-shaped device that is, in practice, observe-only. sg-ready is the same pattern with four coarse levels. the consequences for the model are two, and both are cheap: (1) the planner must never assume commanded = actual — it plans on **measured** power, always (my metered-boiler lesson generalized: feedback is not optional); (2) the loop design is supervisory — openhab is the slow outer loop that moves envelopes, the device is the fast inner loop that acts, and ramp limits + hysteresis exist exactly to keep the outer loop from fighting the inner one. funny enough the grid does the same to us one level up: your §14a is literally an envelope on the house, and my capacity tariff is the economic version of one. it is envelopes all the way down. the device-specific "what can you override and what must you never touch" knowledge then lives where kai put it in his sketch — in the binding that implements the provider/consumer interface and normalizes its device onto one of the classes.

on the consumption profile as a TimeSeries of % — i liked it enough that i did not want to just agree, so i implemented it today. the batch profile now takes an optional normalized shape (`shape="0.1,1.0,0.2"`, fractions of rated power per hour), and the start decision costs each candidate window as the price series weighted by the shape instead of a flat sum — so the start aligns the program's *heavy* hours with the cheap hours. my old ratedW × runtime was exactly the rectangular special case of your idea. one nice detail from the unit test: a shaped dishwasher sometimes waits an hour longer than the rectangular one would, so that its 2 kW heating phase lands on the cheap hour — which is precisely the behaviour you want. one nuance i hit: a *prototype* profile is relative to program start (a shape, so metadata), and it only becomes a real openhab TimeSeries the moment you schedule or start it and it projects onto absolute time. your slp point fits as a prior for the household side — both our systems learn the base load empirically today (you in EM-10, me in my stats), a standard profile would seed that before there is data.

provider side: also implemented — my forecast thing now publishes its full-horizon hourly pv forecast (today + tomorrow, open-meteo with inverter clipping) as TimeSeries future states on a `forecastSeries` channel, next to the aggregate items it always had. which i think proves your pluggable-predictions point: once the provider profile is "a TimeSeries the participant declares", it stops mattering whether the numbers came from open-meteo, your ai model, or someone else's download.

and following my own autonomy conclusion to its end: consumers now take an optional `measure=` item, and a batch program's running-detection trusts measured draw over switch state — a dishwasher's switch may stay ON, its power does not lie.

all of it is in the repo (commit d0165ed; `docs/ENERGY_TAXONOMY.md` updated with autonomy as a cross-cutting property and your profiles idea, credited). when one of the maintainers has a review window i can turn that doc into a github discussion here. no rush from my side.

**Comment by mstormi** [comment]
@stamateviorel I also support what I called generic consumers, on-off switchable these are with extensions like start thresholds, max run and min cooldown times, we probably would want to carry that over, too. Heizstab (heating rod) is just variant of that.
Here's the [user docs on that feature](https://storm.house/docs/#AN-AUS_Verbraucher). (yeah all-German, use a translator)

Plus there's an implementation for cooling devices in weisseWare.rules at the end. A fridge or deep cooler can be operated with intermitting on and off but this ensures it is never switched off longer than a specifyable number of minutes.


Finally for your understanding, my EMS has the key concept of having 4 energy levels (Energieniveau) that represent how available cheap power is. It's calculated from PV surplus and grid price. You've already identified the corresponding flags in tariff rules yourself.
It's intentionally four levels because these can be easily mapped to SGready modes to operate heat pumps as well as to choose one of four evcc modes for charging a car. As you said yourself, this can collapse into two modes, allow/deny effectively, mapped to on and off.
It's documented [here](https://storm.house/docs/#Energieniveaumanagement).

**Comment by kaikreuzer** [comment]
Thanks @stamateviorel for reviving this discussion! Happy to read that you in general confirm my architectural ideas from back then.
And I very much value your and @mstormi practical experience of many years dealing with that topic - and yes, I think we all clearly see the huge complexity and it won't be easy to tackle this. I myself struggle since a while with optimizing my dynamic tariff (Tibber) with my wallbox, battery, PV, heatpump and A/C...

I'm hence also still interested in getting a general solution into openHAB. What didn't exist at the time we last discussed it are LLMs resp. coding assistants that really work well. I'm wondering whether it might be an idea to try out a kind of spec-driven development, since we have a long list of requirements and we are all fairly clear on what we want and have coding assistants work through it to come up with a meaningful implementation. Wdyt, would that be worth a try?

**Comment by stamateviorel** [comment]
markus, i read both docs pages this time (translator did fine) and also the cooling rules at the end of weisseWare — the off-timestamp plus the catch-up time after a restart. good stuff. i carried the whole AN-AUS set over now: priority, per-device threshold, schaltniveau (on at energy level >= N), startklar, min/max runtime, cooldown, and max-off for the fridge — which nicely closed the symmetry, min and max for both on and off. funny detail: your nachholzeit comes for free, after a forced restart the min runtime is the catch-up. and your priority parameter fixed a real bug in mine — without it my allocation order depended on item iteration order.

the level percentiles with the seasonal profiles i did not port yet, that is the config layer on top. everything is in the taxonomy doc, credited.

**Comment by stamateviorel** [comment]
@kaikreuzer honest answer: for me it is not even a hypothesis — this is how the reference was built. the spec was your 2023 design plus this thread, a coding assistant (claude, the co-author is on every commit) did the implementation, i did the reviewing, the testing and the arguing. markus' profile idea from friday morning was running the same afternoon.

the one thing i learned: writing the code is the easy part. what made it trustable was the validation around it — unit tests on the pure logic, then weeks of shadow mode next to my old production rules comparing every decision, and only then live control. it drives my wallbox right now, old pipeline one flag away.

so yes, worth a try. the requirements are basically already collected in docs/ENERGY_TAXONOMY.md in my repo — four device classes, the protections, the energy levels, all from this thread and two production systems. if you want i can turn that into a discussion here and we let the assistants work against it, smallest piece first (the energy: metadata namespace). and your tibber + wallbox + battery + pv + heatpump setup is exactly the class list, so you would be the perfect test case.

**Comment by mstormi** [comment]
We should also prepare for multiple named profiles per device. Think of a washing machine with a fullsize and a low temp program, think of a heat pump that only does hot water once per day in summer (a common heating functionality) but runs all day in winter, think of pv production in summer vs. in winter, and the months inbetween are again different.
In fact, the pv generation window changes by a few minutes every day of the year.
The TimeSeries and the scale factor should be part of the profile.

NOTE: we are open about the time interval this covers. Can be a day for white goods, a week or two for a car and even a whole year for a heat pump.

On seasonal profiles yeah my vision was to have some AI driven layer that optimizes some of the settings such as how long we can switch the heating off before it's getting too uncomfortable by correlating item data like weather report with an inside temp sensor to determine heat loss rate of the house
That learning layer could also adapt/replace the consumption TimeSeries profile from recorded historic data.

**Comment by kaikreuzer** [comment]
> a coding assistant (claude, the co-author is on every commit) did the implementation

RIght, I didn't look at your repo before - now I clearly see Claude having been very active there. Makes absolutely sense.

> so yes, worth a try. the requirements are basically already collected in docs/ENERGY_TAXONOMY.md in my repo 

I thought about gathering such infos in a shared place and enhance it over time. And possibly use a structured process to transform the raw requirements into a PRD and eventually into an implementation (and being able to incrementally repeat things on changes). I was thinking about something like BMAD or Spec-Kit. Having read [this post](https://ranthebuilder.cloud/blog/i-tested-three-spec-driven-ai-tools-here-s-my-honest-take/), OpenSpec might be an interesting candidate as well. Does anyone of you have any experience with one of them (or some similar framework)?

While your implementation is likely a well tested solution already, I believe that our "final" solution should be baked into openhab-core (that's why the discussion is happening here) and not just be "yet another binding". My thoughts on the architecture were that we might introduce new types of add-ons to make the EMS extensible - but we will have to work out to what extend this makes sense.

**Comment by masipila** [comment]
> We should also prepare for multiple named profiles per device. Think of a washing machine with a fullsize and a low temp program, think of a heat pump that only does hot water once per day in summer (a common heating functionality) but runs all day in winter, think of pv production in summer/winter and the months inbetween. The TimeSeries and the scale factor should be part of the profile.

I second the other Markus here. I have also the "summer mode" for my heat pump which does exactly what the other Markus used as an example: it runs the (uninterrupted) domestic hot water cycle once per day. 

> On seasonal profiles yeah my vision was to have some AI driven layer that optimizes some of the settings such as how long we can switch the heating off before it's getting too uncomfortable by correlating item data like weather report with an inside temp sensor to determine heat loss rate of the house That learning layer could also adapt/replace the consumption TimeSeries profile from recorded historic data.

I like this idea. It took me two winter seasons to learn how our house behaves in terms of heating need. The heating need is mostly linear vs. the outdoor temperature but but the passive heating during the springs is a significant factor as well as the big dips in outdoor temperature (here in North the temperature can dip like from 0 to -20 in 12 hours so pre-heating is a very good idea). But this heating need is a forecast TimeSeries in my eyes and there are then different kinds of factors that can contribute to this. Weather forecast is one. I use solar forecast as a proxy for the passive solar heating.

Cheers,
Markus

**Comment by stamateviorel** [comment]
@kaikreuzer i had not used any of the three before your comment. i read the same blog post and picked openspec to try, mostly because its model is current specs plus change deltas, which is pretty much the "repeat on changes" loop you described. spec-kit looked more aimed at greenfield and bmad is a whole methodology on its own.

so instead of an opinion i tried it on this thread: https://github.com/stamateviorel/openhab-ems-spec

i went back and read everything again from january 2023 and put it into openspec changes — your architecture sketch, the backlog @masipila proposed back then and you approved (step 1 of it, storing future timestamped data, actually landed already in 4.1 with timeseries and the forecast strategy), the levels and AN-AUS docs from @mstormi, @lsiepel's demand/provider/engine split, the capacity and peak fee input from @mherwege and @seime, and the named profiles from sunday. it came out as 7 change proposals in 4 waves and every requirement has a source line pointing to the comment or production system it came from. the parts where there is no agreement yet (metadata vs power descriptions vs new add-on types, what goes in core vs addon) i kept as open questions in the design doc, on purpose.

about where it lives: it is under my account only because i had to put it somewhere. if this becomes the shared place i transfer it to the openhab org as-is, or move the content wherever you prefer.

if people want to review something i would start with the smallest one, the participant model — 8 requirements, all from this thread.

**Comment by mstormi** [comment]
A note on TimeSeries in profiles:
we are flexible and open about the time range and value intervals this uses and covers.
Can be a day of range w/ 1 min intervals for white goods, a week or two for a car's usual driving/charging cycle and even a whole year w/ different length intervals for a heat pump.
On the provider side, it can be a day for nailed-down spot market pricing in 15 min intervals, a week for the a-little-less-safe predicted prices and a full year for global power generation.

Note that using different intervals, we can have all of it in a single TimeSeries.

**Comment by stamateviorel** [comment]
good point, and it fits how the core timeseries already works, it is just timestamped entries so the spacing does not need to be uniform. i had it too narrow in the spec, the price part said 60 or 15 min slots. changed it now so a series can be non uniform and even mix intervals in one, fine near term and coarse further out like your predicted-prices and yearly-generation examples, and the profile part says the same for the device curves, a day at 1 min for white goods up to a year with mixed intervals for a heat pump. credited to your comment. on my side the forecast series already does this, future points at whatever resolution the source gives, so it holds up in practice.

**Comment by mstormi** [comment]
Also make sure profile TimeSeries are all read-write and available for updating at any time.
That'll enable us e.g. to initially pre-fill a pv generation profile with a whole year of data based on daily sunshine duration (the window continuously changes every day by some minutes!) but to replace future entries whenever the weather or solar forecast becomes available. 
Plus eventually overwrite even today's pv predictions. Think of an inverter that has a hard limit builtin, very common in Germany and out of reach for openHAB. Larger PV plants can be instructed to dynamically technically limit feed-in at times (§9 EEG, via _Rundsteuerempfänger_).
Would make sense to represent both of these types of upper limit cap in today's generation prediction TimeSeries right away, too.

On the consumption side, you can for example intro a cap to heat pump consumption whenever there's a §14a limit in effect.
Live TimeSeries updates would also allow to adapt consumption predictions for 'dynamic' consumers when a user gets 'deeper' access to data like live consumed power or even more advanced metrics from a heat pump (via Modbus or Cloud interface) while the heat pump itself is still controlled through SGready (which has no concept of power setpoints and measurements).

You can also compute any static physical-technical limits of a device like a maximum battery charge power in there _upfront_ then users don't have to check for each and every of these limits, at least not in each and every part of their code.

This methodology/proceeding will also allow for a good fallback at any time so for the EMS instance to run well even if your internet connection or forecast provider fails for whatever reason, for whatever outage duration.

**Comment by stamateviorel** [comment]
yes, and it is really one idea running through all of it, the prediction series are living not write once. added it now: a generation/consumption/price series is read-write and can be overwritten at any point, past present or future, so you pre-fill a coarse year and let the forecast overwrite the near days as it arrives, and the profile series are the same, updatable live. the inverter or §14a cap fits right in, you write it straight onto today's entries instead of modelling it separately. and the fallback is the real win — with a baseline sitting there the ems keeps planning through an outage, on my side the forecast series is file cached so it already survives a restart or the provider going dark, so that part i can confirm. the sgready-but-measured case i put under the learning layer, control stays coarse but the live modbus power feeds the consumption prediction. all credited. this all lands in the forecast and profile changes, wave 1 is untouched.

and to be straight with everyone: i read and steer this whole thing myself, but the wording of my replies here is polished by claude, the same assistant that did the binding. i go over it but things can slip — small one from my last reply, i said my forecast series already handles the variable resolution, but my source is plain hourly so it is uniform in practice, the honest point is only that the series does not force its own grid. so if something reads too clean or does not add up, please call it out, i would rather have it corrected in the open than let a wrong claim sit in the spec.

**Comment by masipila** [comment]
I quickly skimmed the seven proposal.md pages documented so far.

Regarding the recent discussion about the dynamic nature of the demand side.

How I use my openHAB Spot Price Optimizers for different devices is conceptually this:

1. Heating of the house with ground source heat pump

Tomorrow's demand is calculated for 4 x 6h periods in the afternoon on the day before. This is basically the derived demand described in [1]. And ascdiscussed in the previous comments, this can be stored as a timeseries.

In spring mornings, the solar forecast is checked even though our house does not have PV. If the solar forecast shows that the day will be sunny, the heating demand is lowered for the daytime. In other words, the timeseries gets overwritten for the day part.

(Reason for doing the solar check in the morning and ignoring it on the previous afternoon is that the situation can change during the night).

2. Charging the car

I have an item which I use to select the amount of charging we need during the night between 22 and 06.

Also this can be seen as a derived demand timeseries, the source of the algorithm doing this derivation is much simpler than the heating demand.

Side note: my openHAB Spot Price Optimizer use hours as the unit of measure but in reality the demand is measured in kWh and that is how it should be persisted in the timeseries.

Point of this comment is to share my thinking-out-loud that the concept of derived demand is solid in my eyes. The algorithms producing these derived demands can take whatever they need as inputs, as long as they produce the demand timeseries.

Cheers,
Markus

[1] https://github.com/stamateviorel/openhab-ems-spec/blob/main/openspec/changes/define-forecast-providers/proposal.md




Jul 20, 2026 13:07:23 stamateviorel ***@***.***>:

**Comment by stamateviorel** [comment]
thanks for reading through all seven markus. the heating one is a good example, the morning solar check that lowers the daytime demand is exactly the overwrite case we just added, a prediction series you rewrite as better data lands, and it is nicer coming from a house with no pv than from mine. i put both your examples into the derived-demand change, the heat pump 4x6h with the morning solar overwrite and the simpler night car charging, credited. the kWh point is right too, i fixed the derived demand to say energy not hours, the wave 1 demand declaration was already kWh so they line up now. and yes, the algorithm can take whatever inputs it needs as long as it outputs the demand series, that is the whole reason to keep the engine out of the derivation.

**Comment by mstormi** [comment]
Branching note:
The 2023 thread evolved including a discussion on using EEBus. Kai stated at some stage to separate that from the main discussion.

Nothing happened on that branch part since, but now with AI codegen superpower discussed here that we didn't have back then, can't we have Claude et al implement it, too?
German standards body VDE FNN defined EEBus to be the interface of choice to use to have a smart meter control/dim major consumers *through* an energy management system, so it's an important puzzle piece of the big picture. 

I've opened https://github.com/openhab/openhab-addons/issues/21211

**Comment by stamateviorel** [comment]
quick update in case anyone is still interested here. i took the spec repo further and then actually built it, mostly to find out whether the requirements survive contact with a compiler.

so it is implemented now in a real openhab-core checkout, four bundles, 601 tests, 0 failures. the engine bundle has the participant model, the engine loop, the level classifier and the extension seams, and on top of that the price plane, the forecast plane and the selectable objectives - so it can take a day-ahead price series and a solar forecast and produce a plan, which is the part that was missing when i last posted. checkstyle pmd and spotbugs are zero across all four, javadoc clean, and every dependency is an openhab-core artifact, nothing outside the default set and no http client anywhere, core fetches nothing itself.

it is shadow only and it cannot write to an item at all, and i mean that structurally, the compiled engine bundle does not even import org.openhab.core.items.events and there are tests that fail if somebody adds a write path. anything that touches openhab's own data surfaces sits in separate opt-in companion bundles. @masipila your price tables from 2023 are in there as conformance fixtures, they reproduce exactly, slot by slot.

but the code is not really the point. building it found around ninety places where the corpus was ambiguous or contradictory or just silent, the kind you only hit when you try to type the thing out. priority had no stated direction. "unacknowledged" was never made operational, is 15.999 A an acknowledgement of 16 A. the level scale was named in prose but numbered only inside a test file. the fridge duty cycle and the fuse limit both said "regardless of" and nobody had said which one wins. that whole list is in the repo with what became of each one, and wave 2 threw up another set of the same kind.

i ended up answering them myself, 31 decisions so far, because waiting was not getting anywhere and a spec with eighty open questions is not implementable by anyone. they are recorded as my decisions in a reference implementation, not as agreement here, and every one keeps the options that were rejected written next to the one that was taken. so if you think a decision is wrong you overturn it by pointing at the alternative that is already there, it is one requirement to rewrite and not a re-run of the analysis.

repo is the same one, https://github.com/stamateviorel/openhab-ems-spec, and the offer to move it to the openhab org still stands. i am not going to take it further - grid constraints, the profiles, the ui - until somebody who would have to maintain it has had a look, so if the approach is wrong i would rather hear it now than after another wave.

same disclosure as before, the wording here is polished by claude and the implementation is largely claude's work, i did the reviewing, the testing and the deciding.

**Comment by kaikreuzer** [comment]
Thanks for your update @stamateviorel, I'm definitely still interested (please excuse the slow responses, though...)!
The OpenSpec approach looks nice, I think it gives everything a nicely digestible structure. I'll have a closer look at the details asap.

Wrt the four bundles that you have generated out of it: Where can I find them? In https://github.com/stamateviorel/openhab-binding-emsmanager/tree/main it still looks as the original binding with everything together?

**Comment by stamateviorel** [comment]
sorry, that link was missing because i took the repo down at the end of august. i had decided the thread was not going anywhere and deleted it, which was premature of me. github keeps deleted repos for ninety days so it is back, same commit, nothing lost:

https://github.com/stamateviorel/openhab-core-energy-prototype

the emsmanager one you found is a different thing. that is the binding that actually runs my house, one bundle on purpose, it writes to items and drives the chargers and the battery. the four core bundles cannot do any of that, the engine bundle does not even import org.openhab.core.items.events and there is a test that fails if somebody adds a write path, so they were never going to sit in the same repo.

the four are org.openhab.core.energy for the engine, and three opt-in companions for anything that touches openhab's own data surfaces, .series, .forecast.store and .publish. 608 unit tests and 7 osgi integration tests. the integration ones are the ones worth having, they prove all four resolve and reach active in a real framework and that an item's future prices in persistence come out the other end as a level plan, which is the part i could not prove with unit tests.

PROPOSAL.md in the repo root is written for ten minutes of your time rather than for a reader. the three commands to build and run it, the three things it deliberately refuses to do, what is not done, and whose the 39 decisions are.

i also put the link back in the spec readme. while the repo was down it said the implementation was no longer published, which was true for three weeks and is not now.

i said in august i would not take it further, grid constraints, the profiles, the ui, until somebody who would have to maintain it had looked at it, so nothing has moved since and that still stands. which direction is useful is your call, whether this shape belongs in core at all or as the new add-on types you mentioned, or whether the spec is the part worth keeping and the code was only ever the instrument. the offer to move either repo to the openhab org is still open.

same disclosure as before, the wording here is polished by claude and the implementation is largely claude's work, i did the reviewing, the testing and the deciding.
